### PR TITLE
Vmec adjoint

### PR DESCRIPTION
**Pull request by ejpaul**
This PR includes an implementation of an adjoint-based shape gradient for iota and the magnetic well from vmec equilibria.  The parameter derivatives are computed from the shape gradient using the parameter_derivatives function for any surface type. Included is an example of adjoint-based optimization for a target iota profile.

### CODE REVIEW

File: src/simsopt/mhd/vmec_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        adjoint_epsilon : sets the amplitude of the toroidal
+            current perturbation required for the adjoint solve.
+    """
+
+    def __init__(self, vmec, iota_target, adjoint_epsilon=1.e-1):
+        self.vmec = vmec
+        self.boundary = vmec.boundary
+        self.iota_target = iota_target
+        self.adjoint_epsilon = adjoint_epsilon
+        self.depends_on = ["boundary"]
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/mhd/vmec_diagnostics.py:87]: Is there an advantage to using depends_on = ["vmec"] since I am not computing derivatives with respect to the other dofs?

File: tests/test_files/input.tok [hunk 2]
```diff
@@ -0,0 +1,21 @@
+&INDATA
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply tests/test_files/input.tok]: This is deleted.

File: src/simsopt/geo/surface.py [hunk 3]
```diff
@@ -264,10 +264,51 @@ def aspect_ratio(self):
...
 
         AR = R_major/R_minor
         return AR
+
+    def arclength_poloidal_angle(self):
+        """
+        Computes poloidal angle based on arclenth along magnetic surface at
+        constant phi. This is required for evaluating the adjoint shape gradient
+        for free-boundary calculations.
+        """
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/geo/surface.py:284]: This has been addressed

File: src/simsopt/mhd/vmec.py [hunk 4]
```diff
@@ -589,3 +595,352 @@ def vacuum_well(self):
...
+        self.need_to_run_code = True
+
+        deltaB_dot_B = ((Bx-Bx0)*Bx0 + (By-By0)*By0 + (Bz-Bz0)*Bz0)/delta
+
+        return deltaB_dot_B/(mu0) + perturbation[-1]
+
+    def B_cartesian(self):
+        """
+        Computes cartesian vector components of magnetic field on boundary on grid in
+        toroidal angle and arclength poloidal angle. This is required to
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/mhd/vmec.py]: This has been addressed

File: src/simsopt/geo/surface.py [hunk 5]
```diff
@@ -264,10 +264,51 @@ def aspect_ratio(self):
...
+        theta_arclength = np.zeros_like(gamma[:, :, 0])
+        nphi = len(theta_arclength[:, 0])
+        ntheta = len(theta_arclength[0, :])
+        for iphi in range(nphi):
+            for itheta in range(1, ntheta):
+                dr = np.sqrt((R[iphi, itheta] - R[iphi, itheta-1])**2
+                             + (Z[iphi, itheta] - Z[iphi, itheta-1])**2)
+                theta_arclength[iphi, itheta] = \
+                    theta_arclength[iphi, itheta-1] + dr
+        return theta_arclength
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/geo/surface.py:304]: Previously it was just measuring arclength, so it was unnormalized. I am now normalizing such that it is in [0,1].

File: src/simsopt/geo/surface.py [hunk 6]
```diff
@@ -264,10 +264,51 @@ def aspect_ratio(self):
...
+
+    def interpolate_on_arclength_grid(self, function, theta_evaluate):
+        """
+        Interpolate function onto the theta_evaluate grid in the arclength
+        poloidal angle. This is required for evaluating the adjoint shape gradient
+        for free-boundary calculations.
+        """
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/geo/surface.py:316]: This has been addressed

File: tests/geo/test_surface.py [hunk 7]
```diff
@@ -558,5 +559,58 @@ def test_convert_back(self):
...
+    def test_arclength_poloidal_angle(self):
+        """
+        Compute arclength poloidal angle from circular cross-section tokamak.
+        Check that this matches parameterization angle.
+        """
+        s = get_surface('SurfaceRZFourier', True, mpol=1, ntor=0,
+                        ntheta=200, nphi=5, full=True)
+        s.rc[0, 0] = 5.
+        s.rc[1, 0] = 1.
+        s.zs[1, 0] = 1.
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply tests/geo/test_surface.py]: I have modified this test so that the minor radius is no longer 1. I also added a test with a rotating ellipse boundary and checked that the angle is in [0,1].

File: examples/2_Intermediate/vmec_adjoint.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+surf = vmec.boundary
+surf.all_fixed(True)
+# Slowly increase range of modes in optimization space
+for max_mode in range(3, 6):
+    print(max_mode)
+    surf.fixed_range(mmin=0, mmax=max_mode,
+                     nmin=-max_mode, nmax=max_mode, fixed=False)
+
+    least_squares_serial_solve(prob, grad=True, ftol=1e-12, gtol=1e-12, xtol=1e-12)
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply examples/2_Intermediate/vmec_adjoint.py]: I have modified the example so that it only runs 1 function eval at low resolution if in CI. I have added it to examples/run_vmec_examples. I am now also using scipy.minimize directly rather than the least_squares_serial_solve.

File: src/simsopt/mhd/vmec_diagnostics.py [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def B_cartesian(vmec):
+    """
+    Computes Cartesian vector components of magnetic field on boundary
+    on a grid in the vmec toroidal and poloidal angles. This is
+    required to compute adjoint-based shape gradients.
+    """
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/mhd/vmec_diagnostics.py:34]: This has been addressed

File: src/simsopt/mhd/vmec_diagnostics.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+        :math:`\delta \vec{x}`, the resulting perturbation to the
+        objective function is
+
+        .. math::
+          \delta J(\delta \vec{x}) = \int d^2 x \, G \delta \vec{x} \cdot \vec{n}
+
+        where the integral is over the VMEC boundary surface,
+        :math:`G` is the shape gradient, and :math:`\vec{n}` is the
+        unit normal.
+        """
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/mhd/vmec_diagnostics.py:126]: This has been addressed

File: src/simsopt/mhd/vmec_diagnostics.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/mhd/vmec_diagnostics.py:352]: This has been addressed

File: tests/mhd/test_vmec_diagnostics.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+        ntheta = len(theta1D)
+        theta, phi = np.meshgrid(theta1D, phi1D)
+
+        bmnc = 1.5 * vmec.wout.bmnc[:, -1] - 0.5 * vmec.wout.bmnc[:, -2]
+        xm = vmec.wout.xm_nyq
+        xn = vmec.wout.xn_nyq
+        angle = vmec.wout.xm_nyq[:, None, None] * theta[None, :, :] \
+            - vmec.wout.xn_nyq[:, None, None] * phi[None, :, :]
+        B = np.sum(bmnc[:, None, None] * np.cos(angle), axis=0)
+        self.assertTrue(np.max(np.abs(B**2-B2)) < 1.e-4)
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply tests/mhd/test_vmec_diagnostics.py]: I have changed this to np.testing.assert_allclose

File: src/simsopt/geo/surfaceobjectives.py [hunk 13]
```diff
@@ -244,3 +247,38 @@ def boozer_surface_residual(surface, iota, G, biotsavart, derivatives=0):
...
+    .. math::
+      \frac{\partial f}{\partial \Omega} = \int d^2 x \, G \frac{\partial\vec{x}}{\partial \Omega} \cdot \vec{n},
+
+    where :math:`\Omega` is any parameter of the surface.
+
+    Args:
+        surface: The surface to use for the computation
+        shape_gradient: array-like with same dimensions as angles on the surface,
+            (nphi,ntheta)
+    """
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply src/simsopt/geo/surfaceobjectives.py:279]: This has been addressed

**Comment by landreman** [comment]
Looks great. Could you also merge in master and resolve the merge conflicts?

I'll mention one thing to be aware of. In the vmec_adjoint.py example now, since LeastSquaresProblem was removed, I'm pretty sure the fixed dofs are not actually held fixed. This is because set_dofs() and get_dofs() set and return all the dofs, even fixed ones. This means "max_modes" is not respected. To pick out just the non-fixed dofs, I think after each call to surf.fixed_range() in the loop you can create an object myfunc=simsopt._core.dofs.Dofs(obj), where obj is the IotaTargetMetric already defined. Then I think myfunc.f and myfunc.jac give handles to the objective function and gradient for only the non-fixed dofs. Since Dofs() works with vector-valued objectives, you probably need to pick out the first element of f and squash a dimension of jac.
(This will likely change when Bharat gets the new graph system working...)

**Comment by ejpaul** [comment]
I have modified the example as you suggest and merged in master.
